### PR TITLE
Too many submissions penalty

### LINKED ISSUE #487: make test case for "too many submissions" penalty
And allow customization of the # of penalty free submissions, and the amount of penalty.

### PR DESCRIPTION
**Pull request by bmcutler**
closes #487

The badge & title positions are sloppy now.  Would be better to ensure that the test titles are lined up regardless of the presence or size of the badge.

I'll try to attach images...
screen shot 2016-08-14 at 3 49 04 pm
screen shot 2016-08-14 at 3 48 53 pm

### CODE REVIEW

**Comment by bmcutler** [comment]
The images above are from cpp_hidden_tests with various file uploads.  

Yes, sometimes the badge isn't there (if its an extra credit test and the score is zero, no badge.  or if its a penalty test and the score is zero, no badge).

**Comment by MasterOdin** [comment]
Moved discussion of badges display to #494 so I can work on it after this is merged with master.

File: site/app/models/GradeableTestcase.php [hunk 1]
```diff
@@ -150,6 +164,7 @@ public function getAutochecks() {
     }
     
     public function hasDetails() {
+        if ($this->isHidden()) return false;
         return count($this->autochecks) > 0 || $this->hasCompilationOutput() || $this->hasExecuteLog();
```
Comment (MasterOdin) [inline site/app/models/GradeableTestcase.php]: I think it'd be better to just make this:
```
return (count($this->autochecks) > 0) || $this->hasCompilationOutput() || $this->hasExecuteLog()) && !$this->isHidden()
```

I really dislike one line if statements

File: site/app/views/submission/HomeworkView.php [hunk 2]
```diff
@@ -445,21 +445,38 @@ public function showGradeable($gradeable, $gradeable_select, $days_late) {
...
-                                $background = "";
-                                if ($testcase->getPointsAwarded() >= $testcase->getPoints()) {
-                                    $background = "green-background";
+                              $background = "";
+                              if ($testcase->isExtraCredit()) {
+                                if ($testcase->getPointsAwarded() > 0) {
+                                  $background = "green-background";
+                                  $return .= <<<HTML
+                <span class="badge {$background}"> &nbsp; +{$testcase->getPointsAwarded()} &nbsp; </span>
+HTML;
```
Comment (MasterOdin) [inline site/app/views/submission/HomeworkView.php]: This span should be after this inner most if statement

Comment (bmcutler) [inline site/app/views/submission/HomeworkView.php]: Yes, it is intentional to not have a badge if you don't earn the extra credit.
Yes, it is intentional to not put how many points are possible.
I feel strongly about both these decisions based on my use of extra credit and the appearance when I tested sample (what I thought were typical) submissions.
I'd like to leave this.  (can be re-discussed in future).

File: site/app/models/GradeableTestcase.php (hunk 1 above)
Comment (bmcutler) [inline site/app/models/GradeableTestcase.php]: I think the long line of checks with ands, ors, and nots is harder to read and maintain.
Anyhow, I've cleaned up some dead code so it's your kind of one-liner now.
